programs/bpf_loader/src/syscalls/cpi.rs

```diff
@@ -1244,29 +1524,112 @@ fn update_caller_account(
+            //
+            // Therefore we temporarily configure the realloc region as writable
+            // then set it back to whatever state it had.
+            let _guard = {
+                let realloc_region = caller_account
+                    .realloc_region(memory_mapping, is_loader_deprecated)?
+                    .unwrap(); // unwrapping here is fine, we asserted !is_loader_deprecated
+                let state = realloc_region.state.replace(MemoryState::Writable);
+
+                ReallocStateDropGuard {
+                    realloc_region,
+                    state,
+                }
+            };
+
+            translate_slice_mut::<u8>(
+                memory_mapping,
+                caller_account
+                    .vm_data_addr
+                    .saturating_add(caller_account.original_data_len as u64),
+                realloc_bytes_used as u64,
+                invoke_context.get_check_aligned(),
+                invoke_context.get_check_size(),
+            )?
+        };
         let from_slice = callee_account
             .get_data()
             .get(caller_account.original_data_len..post_len)
             .ok_or(SyscallError::InvalidLength)?;
+        if to_slice.len() != from_slice.len() {
+            return Err(Box::new(InstructionError::AccountDataTooSmall));
+        }
         to_slice.copy_from_slice(from_slice);
     }
 
     Ok(())
 }
 
+fn account_data_region<'a>(
+    memory_mapping: &'a MemoryMapping<'_>,
+    vm_data_addr: u64,
+    original_data_len: usize,
+) -> Result<Option<&'a MemoryRegion>, Error> {
+    if original_data_len > 0 {
+        // We can trust vm_data_addr to point to the correct region because we
+        // enforce that in CallerAccount::from_(sol_)account_info.
+        let region = memory_mapping.region(AccessType::Load, vm_data_addr)?;
+        // vm_data_addr must always point to the beginning of the region
+        debug_assert_eq!(region.vm_addr, vm_data_addr);
+        return Ok(Some(region));
+    }
+
+    Ok(None)
+}
+
+fn account_realloc_region<'a>(
+    memory_mapping: &'a MemoryMapping<'_>,
+    vm_data_addr: u64,
+    original_data_len: usize,
+    is_loader_deprecated: bool,
+) -> Result<Option<&'a MemoryRegion>, Error> {
+    if is_loader_deprecated {
+        return Ok(None);
+    }
+
+    let realloc_vm_addr = vm_data_addr.saturating_add(original_data_len as u64);
+    let realloc_region = memory_mapping.region(AccessType::Load, realloc_vm_addr)?;
+    debug_assert_eq!(realloc_region.vm_addr, realloc_vm_addr);
+    debug_assert!((MAX_PERMITTED_DATA_INCREASE
+        ..MAX_PERMITTED_DATA_INCREASE.saturating_add(BPF_ALIGN_OF_U128))
+        .contains(&(realloc_region.len as usize)));
+    Ok(Some(realloc_region))
+}
+
+struct ReallocStateDropGuard<'a> {
+    realloc_region: &'a MemoryRegion,
+    state: MemoryState,
+}
+
+impl<'a> Drop for ReallocStateDropGuard<'a> {
+    fn drop(&mut self) {
+        self.realloc_region.state.set(self.state);
```

Comment:
how about using `scopeguard::defer!()` not to clutter these business logics?

Reply:
I've done this because I don't feel strongly about it, but I personally think that the explicit guard code used before was _easier_ to understand since it was explicit and it was like 5 lines of code, which in my mind always beats introducing a 3rd party dependency 😋